core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -101,27 +106,97 @@
   private final boolean inHeap;
 
   /**
+   * The size of memory granted to each consumer.
+   */
+  private final HashMap<MemoryConsumer, Long> consumers;
+
+  /**
    * Construct a new TaskMemoryManager.
    */
   public TaskMemoryManager(MemoryManager memoryManager, long taskAttemptId) {
     this.inHeap = memoryManager.tungstenMemoryIsAllocatedInHeap();
     this.memoryManager = memoryManager;
     this.taskAttemptId = taskAttemptId;
+    this.consumers = new HashMap<>();
   }
 
   /**
-   * Acquire N bytes of memory for execution, evicting cached blocks if necessary.
+   * Acquire N bytes of memory for a consumer. If there is no enough memory, it will call
+   * spill() of consumers to release more memory.
+   *
    * @return number of bytes successfully granted (<= N).
    */
-  public long acquireExecutionMemory(long size) {
-    return memoryManager.acquireExecutionMemory(size, taskAttemptId);
+  public long acquireExecutionMemory(long size, MemoryConsumer consumer) throws IOException {
+    assert(size >= 0);
+    synchronized (this) {
+      long got = memoryManager.acquireExecutionMemory(size, taskAttemptId);
+
+      // call spill() on itself to release some memory
+      if (got < size && consumer != null) {
+        consumer.spill(size - got, consumer);
+        got += memoryManager.acquireExecutionMemory(size - got, taskAttemptId);
+      }
+
+      if (got < size) {
+        long needed = size - got;
+        // call spill() on other consumers to release memory
+        for (MemoryConsumer c: consumers.keySet()) {
+          if (c != null && c != consumer) {
+            needed -= c.spill(size - got, consumer);
+            if (needed < 0) {
+              break;
+            }
+          }
+        }
+        got += memoryManager.acquireExecutionMemory(size - got, taskAttemptId);
```

Comment:
This function does not garantee that `got` will be equal to `size`, after call spill().